src/content_main.js

```diff
@@ -171,42 +162,134 @@ async function storeRecentTimeSpent(recentTimeSpent) {
+			// Therefore it is necessary to use handleValidationAndSetExtensionIcon which checks the
+			// browser storage to see if the payment pointer has been recently validated.
+			handleValidationAndSetExtensionIcon(paymentPointerOnPage);
+			cachedPaymentPointer = paymentPointerOnPage;
+		}
+	}, NEW_PAYMENT_POINTER_CHECK_RATE_MS);
+}
+
+/**
+ * Ensures that the extension icon is updated to the correct state when switching between tabs.
+ */
+function handleUpdatingExtensionIconOnTabVisibilityChange() {
+	document.addEventListener('visibilitychange', (event) => {
+		const tabIsVisible = !document.hidden;
+
+		if (tabIsVisible) {
+			const paymentPointerOnPage = getPaymentPointerFromPage();
+			if (paymentPointerOnPage === null) {
+				setExtensionIconMonetizationState(false);
+			} else {
+				handleValidationAndSetExtensionIcon(paymentPointerOnPage);
+			}
+		}
+	});
+}
+
+/**
+ * Given a payment pointer, if it is valid, this function sets the extension icon to its
+ * monetized state and stores the payment pointer in browser storage,
+ * otherwise the extension icon is set to its unmonetized state.
+ *
+ * @param {string} paymentPointer the payment pointer to validate.
+ * @returns {Promise<>} a promise that resolves when validation is finished.
+ */
+async function handleValidationAndSetExtensionIcon(paymentPointer) {
+	if (await isPaymentPointerRecentlyValidated(paymentPointer)) {
+		setExtensionIconMonetizationState(true);
+	} else {
+		// validate if not validated recently
 		if (await isPaymentPointerValid(paymentPointer)) {
-			isValid = true;
+			// Store the payment pointer and the UTC timestamp for the time it was validated (now).
+			await storeDataIntoAkitaFormat({
+				paymentPointer: paymentPointer,
+				validationTimestamp: Date.now()
+			}, AKITA_DATA_TYPE.PAYMENT);
+			setExtensionIconMonetizationState(true);
+		} else {
+			setExtensionIconMonetizationState(false);
 		}
 	}
+}
 
-	return {
-		isValid,
-		paymentPointer
-	};
+/**
+ * Checks the browser storage to see if Akita has validated the given payment pointer and stored a
+ * validation timestamp for it recently.
+ *
+ * @param {string} paymentPointer the payment pointer to check.
+ * @returns {Promise<boolean>} resolves to true if there is a timestamp retrieved from storage and
+ * the timestamp is recent (less than PAYMENT_POINTER_VALIDATION_RATE_MS from present time).
+ * Resolves false otherwise.
+ */
+async function isPaymentPointerRecentlyValidated(paymentPointer) {
+	const originData = await loadOriginData(window.location.origin);
+	let validationTimestamp = null;
+	if (originData && originData.paymentPointerMap[paymentPointer]) {
+		// get validationTimestamp from originData under the given payment pointer.
+		validationTimestamp =
+			originData.paymentPointerMap[paymentPointer].validationTimestamp;
+	}
+	if (validationTimestamp === null) {
+		return false;
+	}
+
+	// check if validationTimestamp is recent.
```

Comment:
Start comment with capital and do not use period at end if it's a single line/sentence